pkg/gds/admin/v2/csurf_test.go

```diff
@@ -0,0 +1,175 @@
+)
+
+func TestDoubleCookies(t *testing.T) {
+	// Test both the DoubleCookie middleware and the SetDoubleCookieTokens handler.
+	router := gin.New()
+
+	// Add login route that sets the cookies
+	router.GET("/login", func(c *gin.Context) {
+		err := admin.SetDoubleCookieTokens(c, time.Now().Add(time.Minute*10).Unix())
+		require.NoError(t, err)
+		c.JSON(http.StatusOK, gin.H{"success": true})
+	})
+
+	// Add request route that requires double cookie submit
+	router.POST("/action", admin.DoubleCookie(), func(c *gin.Context) {
+		c.JSON(http.StatusCreated, gin.H{"success": true})
+	})
+
+	// Create an http client with a cookie jar
+	jar, err := cookiejar.New(nil)
+	require.NoError(t, err)
+	client := http.Client{
+		Jar: jar,
+	}
+
+	// Create the test server with the CSRF protected router
+	server := httptest.NewServer(router)
+
+	// Attempt to make a request to the server that is not CSRF protected
+	req, err := http.NewRequest(http.MethodPost, server.URL+"/action", nil)
+	require.NoError(t, err)
+
+	// Ensure the request is Forbidden
+	rep, err := client.Do(req)
+	require.NoError(t, err)
+	require.Equal(t, http.StatusForbidden, rep.StatusCode)
+
+	// Check the data in the response
+	data, err := readJSON(rep)
+	require.NoError(t, err)
+	require.Contains(t, data, "error")
+	require.Contains(t, data, "success")
+	require.Equal(t, admin.ErrCSRFVerification.Error(), data["error"].(string))
+	require.False(t, data["success"].(bool))
+
+	// Login and set the cookies
+	req, err = http.NewRequest(http.MethodGet, server.URL+"/login", nil)
+	require.NoError(t, err)
+
+	// Ensure the request is Ok
+	rep, err = client.Do(req)
+	require.NoError(t, err)
+	require.Equal(t, http.StatusOK, rep.StatusCode)
+
+	// Check that we're getting two cookies back in the response
+	cookies := rep.Cookies()
+	require.Len(t, cookies, 2)
+
+	// Check the data in the response
+	data, err = readJSON(rep)
+	require.NoError(t, err)
+	require.NotContains(t, data, "error")
+	require.Contains(t, data, "success")
+	require.True(t, data["success"].(bool))
+
+	// TESTING HACK: add the secure cookies to the client cookie jar even though this
+	// is not TLS. We're interested in the DoubleCookie middleware, not whether or not
+	// the http client does the right thing with cookies.
+	ep, _ := url.Parse(server.URL)
+	for _, cookie := range cookies {
+		cookie.Secure = false
+	}
+	client.Jar.SetCookies(ep, cookies)
+
+	// Attempt to send a request with the cookies but no X-CSRF-TOKEN header
+	req, err = http.NewRequest(http.MethodPost, server.URL+"/action", nil)
+	require.NoError(t, err)
+
+	// Ensure the request is Forbidden
+	rep, err = client.Do(req)
+	require.NoError(t, err)
+	require.Equal(t, http.StatusForbidden, rep.StatusCode)
+
+	// Check the data in the response
+	data, err = readJSON(rep)
+	require.NoError(t, err)
+	require.Contains(t, data, "error")
+	require.Contains(t, data, "success")
+	require.Equal(t, admin.ErrCSRFVerification.Error(), data["error"].(string))
```

Comment:
Is it important that this error is indistinguishable from the one on line 58? E.g. we want them to know they got it wrong, but not how they got it wrong?

Reply:
This is definitely a question I go back and forth on. I think that the best practice is that you do want the error to be only as informative as possible for the front end to make a decision (e.g. redirect to login in this case); but not so informative that an attacker use the messages to analyze for vulnerabilities. The idea is that the the logs contain debugging information not the response itself. 

However, better front end errors help debug on the client side and also help find the log messages that may contain more information. The question is what does the front-end do? Does it just console.log the error in the front end or does it display the text to the user? 

Potentially another happy hour discussion?